src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -170,6 +253,129 @@ public static bool CursorVisible
             }
         }
 
+        // TODO: It's quite expensive to use the request/response protocol each time CursorLeft/Top is accessed.
+        // We should be able to (mostly) track the position of the cursor in locals, doing the request/response infrequently.
+
+        public static int CursorLeft
+        {
+            get
+            {
+                int left, top;
+                GetCursorPosition(out left, out top);
+                return left;
+            }
+        }
+
+        public static int CursorTop
+        {
+            get
+            {
+                int left, top;
+                GetCursorPosition(out left, out top);
+                return top;
+            }
+        }
+
+        /// <summary>Gets the current cursor position.  This involves both writing to stdout and reading stdin.</summary>
+        private static unsafe void GetCursorPosition(out int left, out int top)
+        {
+            left = top = 0;
+
+            // Getting the cursor position involves both writing out a request string and
+            // parsing a response string from the terminal.  So if anything is redirected, bail.
+            if (Console.IsInputRedirected || Console.IsOutputRedirected)
+                return;
+
+            // Get the cursor position request format string.
+            string cpr = TerminalBasicInfo.Instance.CursorPositionRequestFormat;
+            if (string.IsNullOrEmpty(cpr))
+                return;
+
+            // Synchronize with all other stdin readers.  We need to do this in case multiple threads are
+            // trying to read/write concurrently, and to minimize the chances of resulting conflicts.
+            // This does mean that Console.get_CursorLeft/Top can't be used concurrently Console.Read*, etc.;
+            // attempting to do so will block one of them until the other completes, but in doing so we prevent
+            // one thread's get_CursorLeft/Top from providing input to the other's Console.Read*.
+            lock (StdInReader) 
+            {
+                // Write out the cursor position request.
+                WriteStdoutAnsiString(cpr);
+
+                // Read the response.  There's a race condition here if the user is typing,
+                // or if other threads are accessing the console; there's relatively little
+                // we can do about that, but we try not to lose any data.
+                StdInStreamReader r = StdInReader.Inner;
+                const int BufferSize = 1024;
+                byte* bytes = stackalloc byte[BufferSize];
+
+                int bytesRead = 0, i = 0;
+
+                // Response expected in the form "\ESC[row;colR".  However, user typing concurrently
```

Comment:
It is scary, but it's the nature of the beast: the protocol involves writing to stdout, at which point data gets put on stdin for reading.

Your understanding is correct, in that any other input that came in at the same time could be misintepreted.  We might also skip over such input.  I tried an approach where I put such data back into the ReadKey internal buffer so that it would subsequently be consumed, but then we potentially also end up with garbage data in the ReadKey buffer.

The window is small, but it's there.  It's most reproably if you hold down a key that generates an escape sequence, e.g. an arrow key, and have code that processes an arrow key by checking for the current position.  Then you end up having data showing up on stdin that's either from the escape sequence generated from pressing the arrow key or the escape sequence generated as a response to the current cursor position request.

I wouldn't expect much if any real input to get lost.  Any that did would be coming from someone typing, and it would likely be due to holding down a character key to have it repeating, in which case losing one or more isn't terrible as it's repeating.  Such concerns would also be mitigated by not checking for the current cursor position on every key request, which is what I was doing in my sample apps stressing this.  I have a TODO in the PR, but we could look into having some fields attempt to track the current cursor position, e.g. get the values once at the start of the app, and then based on the keys that are read, try to update those values to be kept in sync with the real cursor position rather than reading it every time.  That has its own risks, though.